src/NuGet.Core/NuGet.Build.Tasks/StaticGraphRestoreTaskBase.cs

```diff
@@ -0,0 +1,273 @@
+
+                        process.Start();
+
+                        WriteArguments(process.StandardInput);
+
+                        process.BeginOutputReadLine();
+
+                        semaphore.Wait(_cancellationTokenSource.Token);
+
+                        if (!process.HasExited)
+                        {
+                            try
+                            {
+                                process.Kill();
+                            }
+                            catch (InvalidOperationException)
+                            {
+                                // The process may have exited, in this case ignore the exception
+                            }
+                        }
+                    }
+                    catch (Exception e) when (
+                        e is OperationCanceledException
+                        || (e is AggregateException aggregateException && aggregateException.InnerException is OperationCanceledException))
+                    {
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Log.LogErrorFromException(e);
+            }
+
+            return !Log.HasLoggedErrors;
+        }
+
+        /// <summary>
+        /// Gets the command-line arguments to use when launching the process that executes the restore.
+        /// </summary>
+        internal IEnumerable<string> GetCommandLineArguments(string msbuildBinPath)
+        {
+#if IS_CORECLR
+
+            yield return Path.Combine(ThisAssemblyLazy.Value.DirectoryName, Path.ChangeExtension(ThisAssemblyLazy.Value.Name, ".Console.dll"));
+
+            yield return Path.Combine(msbuildBinPath, "MSBuild.dll");
+#else
+            yield return Path.Combine(msbuildBinPath, "MSBuild.exe");
+#endif
+            yield return IsSolutionPathDefined ? SolutionPath : ProjectFullPath;
+        }
+
+        /// <summary>
+        /// Enumerates a list of global properties for the current MSBuild instance.
+        /// </summary>
+        /// <returns>A <see cref="Dictionary{TKey, TValue}" /> containing global properties.</returns>
+        internal virtual Dictionary<string, string> GetGlobalProperties()
+        {
+            IReadOnlyDictionary<string, string> globalProperties = null;
+
+#if IS_CORECLR
+            // MSBuild 16.5 and above has a method to get the global properties, older versions do not
+            if (BuildEngine is IBuildEngine6 buildEngine6)
+            {
+                globalProperties = buildEngine6.GetGlobalProperties();
+            }
+#else
+            // MSBuild 16.5 added a new interface, IBuildEngine6, which has a GetGlobalProperties() method.  However, we compile against
+            // Microsoft.Build.Framework version 4.0 when targeting .NET Framework, so reflection is required since type checking
+            // can't be done at compile time
```

Comment:
Out of scope for this PR, but why don't we compile this project with Microsoft.Build.Framework 17.x?  This project/package doesn't need to support multiple versions of MSBuild like nuget.exe does.

Reply:
I'm assuming its because we ILMerge this assembly into NuGet.exe?  Does that sound right?

Reply:
Yes, you're right. When I wrote the comment, I thought we could use different package versions in different projects, but yes, since nuget.exe is ilmerged we can't 😭